contracts/payment/SponsorPayment.sol

```diff
@@ -0,0 +1,519 @@
+    }
+
+    /**
+     * @dev Updates owner percentage value
+     * @param ownerPercentage Amount between 0 and 100 representing the owner percentage
+     */
+    function updateOwnerPercentage(
+        uint8 ownerPercentage
+    ) external onlyOwner validPercentValue(ownerPercentage) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        $.ownerPercentage = ownerPercentage;
+    }
+
+    function _tryRecoverSigner(
+        bytes32 digest,
+        bytes memory signature
+    ) private pure returns (address) {
+        (address signer, ECDSA.RecoverError err, ) = digest.tryRecover(signature);
+        if (err != ECDSA.RecoverError.NoError) revert InvalidPaymentClaim("Invalid signature");
+
+        return signer;
+    }
+
+    function recoverSponsorPaymentSigner(
+        SponsorPaymentInfo memory payment,
+        bytes memory signature
+    ) public view returns (address) {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    PAYMENT_CLAIM_DATA_TYPE_HASH,
+                    payment.recipient,
+                    payment.amount,
+                    payment.expiration,
+                    payment.nonce,
+                    keccak256(payment.metadata)
+                )
+            )
+        );
+
+        return _tryRecoverSigner(digest, signature);
+    }
+
+    function recoverSponsorPaymentSignerERC20(
+        ERC20SponsorPaymentInfo memory payment,
+        bytes memory signature
+    ) public view returns (address) {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    ERC20_PAYMENT_CLAIM_DATA_TYPE_HASH,
+                    payment.recipient,
+                    payment.amount,
+                    payment.token,
+                    payment.expiration,
+                    payment.nonce,
+                    keccak256(payment.metadata)
+                )
+            )
+        );
+
+        return _tryRecoverSigner(digest, signature);
+    }
+
+    /**
+     * @notice Deposits Ether or tokens as a sponsor.
+     */
+    function deposit() external payable nonReentrant {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+
+        if (msg.value == 0) revert InvalidDeposit("Invalid value amount");
+        $.balances[_msgSender()][address(0)] += msg.value;
+
+        emit Deposit(_msgSender(), msg.value);
+    }
+
+    /**
+     * @notice Deposits Ether or tokens as a sponsor.
+     * @param amount The amount to deposit
+     * @param token The address of the token (use address(0) for Ether)
+     */
+    function depositERC20(uint256 amount, address token) external nonReentrant validToken(token) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        if (amount == 0) revert InvalidDeposit("Invalid token amount");
+        IERC20(token).safeTransferFrom(_msgSender(), address(this), amount);
```

Comment:
is allowance in token contract is given sepatelly?

Reply:
You can see the interaction with SC by reviewing unit tests